RecoTracker/TkTrackingRegions/plugins/AreaSeededTrackingRegionsProducer.h

```diff
@@ -0,0 +1,118 @@
+#include "DataFormats/Math/interface/PtEtaPhiMass.h"
+
+#include "VertexBeamspotOrigins.h"
+#include "AreaSeededTrackingRegionsBuilder.h"
+
+#include <array>
+#include <limits>
+
+/** class AreaSeededTrackingRegionsProducer
+ *
+ * eta-phi TrackingRegions producer in directions defined by z-phi area-based objects of interest
+ * from the "input" parameters.
+ *
+ * Four operational modes are supported ("mode" parameter):
+ *
+ *   BeamSpotFixed:
+ *     origin is defined by the beam spot
+ *     z-half-length is defined by a fixed zErrorBeamSpot parameter
+ *   BeamSpotSigma:
+ *     origin is defined by the beam spot
+ *     z-half-length is defined by nSigmaZBeamSpot * beamSpot.sigmaZ
+ *   VerticesFixed:
+ *     origins are defined by vertices from VertexCollection (use maximum MaxNVertices of them)
+ *     z-half-length is defined by a fixed zErrorVertex parameter
+ *   VerticesSigma:
+ *     origins are defined by vertices from VertexCollection (use maximum MaxNVertices of them)
+ *     z-half-length is defined by nSigmaZVertex * vetex.zError
+ *
+ *   If, while using one of the "Vertices" modes, there's no vertices in an event, we fall back into
+ *   either BeamSpotSigma or BeamSpotFixed mode, depending on the positiveness of nSigmaZBeamSpot.
+ *
+ */
+class AreaSeededTrackingRegionsProducer {
+public:
+
+  typedef enum {BEAM_SPOT_FIXED, BEAM_SPOT_SIGMA, VERTICES_FIXED, VERTICES_SIGMA } Mode;
+
+  AreaSeededTrackingRegionsProducer(const edm::ParameterSet& conf, edm::ConsumesCollector&& iC):
+    m_origins(conf.getParameter<edm::ParameterSet>("RegionPSet"), iC),
+    m_builder(conf.getParameter<edm::ParameterSet>("RegionPSet"), iC)
+  {
+    edm::ParameterSet regPSet = conf.getParameter<edm::ParameterSet>("RegionPSet");
+    for(const auto& area: regPSet.getParameter<std::vector<edm::ParameterSet> >("areas")) {
+      m_areas.emplace_back(area.getParameter<double>("rmin"),
+                           area.getParameter<double>("rmax"),
+                           area.getParameter<double>("phimin"),
+                           area.getParameter<double>("phimax"),
+                           area.getParameter<double>("zmin"),
+                           area.getParameter<double>("zmax"));
+    }
+    if(m_areas.empty())
+      throw cms::Exception("Configuration") << "Empty 'areas' parameter.";
+  }
+  
+  ~AreaSeededTrackingRegionsProducer() = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+
+    edm::ParameterSetDescription descAreas;
+    descAreas.add<double>("rmin", 0.0);
+    descAreas.add<double>("rmax", 0.0);
+    descAreas.add<double>("zmin", 0.0);
+    descAreas.add<double>("zmax", 0.0);
+    descAreas.add<double>("phimin", 0.0);
+    descAreas.add<double>("phimax", 0.0);
+    std::vector<edm::ParameterSet> vDefaults;
+    /*
+    edm::ParameterSet vDefaults1;
+    vDefaults1.addParameter<double>("rmin", 0.0);
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

Reply:
No, will remove.